File: Employee_Importance_690.java
//Time Complexity : O(V+E)
//Space Complexity : O(max(V, E))
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

import java.util.HashMap;
import java.util.List;
import java.util.LinkedList;
import java.util.Queue;

class Employee
{
	public int id;
	public int importance;
	public List<Employee>  subordinates;
}
class Employee_Importance_690 {
    public int getImportance(List<Employee> employees, int id) {
    	if(employees == null || employees.size() == 0)
    		return 0;
    	HashMap<Integer, Employee> map = new HashMap<>();
    	for(Employee e : employees)
    	{
    		map.put(e.id,  e);
    	}
    	Queue<Integer> q = new LinkedList<>();
    	q.add(id);
    	int result = 0;
    	while(!q.isEmpty())
    	{
    		int eid = q.poll();
    		Employee e = map.get(eid);
    		result += e.importance;
    		for(int sid : e.subordinates)
    		{
    			q.add(sid);
    		}
    	}
    	return result;
    }
}

File: employeeImportance_690.py
# // Time Complexity : O(n), n is the number of employees
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

# // Your code here along with comments explaining your approach: 

"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        
        hMap = {}
        
        #add employees to the hashmap-> O(n)
        for emp in employees:
            hMap[emp.id] = emp
        
        output = 0 
        
        queue = []
        #Implementing BFS
        #add root to queue
        queue.append(id)
        
        while queue:
            front = queue.pop(0)
            output += hMap[front].importance
            #go through all the subordinates of the employee
            for sub in hMap[front].subordinates:
                queue.append(sub)
                
        return output
            
        
        
        
        